src/tracing/trace_flamegraph.h

```diff
@@ -0,0 +1,47 @@
+/*
+   +----------------------------------------------------------------------+
+   | Xdebug                                                               |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2002-2020 Derick Rethans                               |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 1.01 of the Xdebug license,   |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available at through the world-wide-web at                           |
+   | https://xdebug.org/license.php                                       |
+   | If you did not receive a copy of the Xdebug license and are unable   |
+   | to obtain it through the world-wide-web, please send a note to       |
+   | derick@xdebug.org so we can mail you a copy immediately.             |
+   +----------------------------------------------------------------------+
+ */
+#ifndef XDEBUG_TRACE_FLAMEGRAPH_H
+#define XDEBUG_TRACE_FLAMEGRAPH_H
+
+#define XDEBUG_TRACE_FLAMEGRAPH_COST_FACTOR 100000;
+
+#include "tracing_private.h"
+
+typedef struct _flamegraph_stack_item
+{
+	char *prefix;
+	int value;
+	struct _flamegraph_stack_item *next;
+} flamegraph_stack_item;
+
+/* Implemented as a linked-list, but size will never be more than the current
+   stack depth, so I guess this will not hurt performances too badly. */
```

Comment:
As I mentioned earlier, I think this should use Xdebug's vector "type". I'm also wondering whether it makes sense to keep a separate stack here, and whether this information can't be stored in a flamegraph-trace specific part of Xdebug's general stack, such as I also do for the [profiler](https://github.com/xdebug/xdebug/blob/master/src/profiler/profiler.c#L272).

Storing it as part of that, would also mean that you don't have to take care of fibres (much), as Xdebug's stack handles that - although you'd still need to check to prefix things with the right fiber "id". That probably means we need to set a new `XG()` variable when a fiber switch is [signalled](https://github.com/xdebug/xdebug/blob/master/src/base/base.c#L1076). There certainly needs to be a test for fibers, which you can adopt from https://github.com/xdebug/xdebug/blob/master/tests/debugger/fiber-001.inc

Reply:
> As I mentioned earlier, I think this should use Xdebug's vector "type". I'm also wondering whether it makes sense to keep a separate stack here

As I stated upper, the only reason why I did kept a separate stack is in order to keep the precomputed line prefix. Flamegraph output lines are like this:

`{main};func1;func2_called_by_func1;... COST`

Which means that in order to build this string, it is required to traverse the current full stack. By using a custom pre-computed prefix for each item in the custom stack, I reduce this operation to only "fetch parent prefix then append current function name" instead of "for each parent in the stack starting from root, append name".

Reply:
> Storing it as part of that, would also mean that you don't have to take care of fibres (much), as Xdebug's stack handles that - although you'd still need to check to prefix things with the right fiber "id". 

OK so when we enter a function call, it might be at its rightful position in the stack, OR if a fiber just got interrupted a new stack, if I understand it ?

Indeed, I need to be able to detect those use cases to handle those properly, and rebuild a complete function prefix when that happens. This also mean that in order to compute the self function cost, I cannot rely upon the previous function I entered, I also need to ensure it's in the same fiber stack.

But if I understand this well, potentially all tracing formats are compromised by this behavior ?

So what I need, I guess, if I don't want to recompute the string prefix and re-traverse the complete stack for every function, is:

 - to be able to detect fiber stack change (more generally, any stack change),
 - store not only a single custom stack, but a hashmap of stacks where each is identifier by the fiber id (more generally, the stack id).

And then, in order to compute the correct 'self' cost, if I implement what I describe upper, I can keep my current algorithm if my custom stack is indexed using the current stack/fiber id.

In the opposite side, if I choose to revert the custom stack stuff, and go back to my original implementation (i.e. fully traverse the stack to compute the prefix) for each function enter/exit, then:

 - we go back to square one and possible terrible performance,
 - when I reach the exit hook, I still need to add the current function inclusive cost to a counter for its parent, so I do need to store this information somewhere, in a custom<function_nr, children_cost> hashmap.

In both case I need an hashmap implementation (or anyother searchable collection if that matters, I don't think having a real hashmap would be a huge benefit here, because the maximum number of elements in it will be the stack depth anyway).

Reply:
I need some time to think about that and explore possibilities.

Reply:
@derickr I did as much as I could in order to apply (almost all) your suggestions. I think the code looks now nice:

 - now use the current stack vector to find the parent function stack entry number in both enter and exit hook, it works nicely,
 - it never traverses the stack, there is no need as I implemented it (I hope),
 - uses the xdebug_hash type for storing function costs temporarily when children are running, I hope it's fast enough.

This version is supposed to be fibers-safe, I added a fibers test, nevertheless I have still two problems:

 - I fear that using a xdebug_hash may be slow, I mean it's going to add an item, compute some values from it, and remove the same item, lookup twice, for each PHP function call (seems overkill).
 - This exact line: https://github.com/xdebug/xdebug/pull/761/files#diff-40553f5cbf2846e70449d253e1cfe8694302f540b54e40c5a2ae70834b4c1959R57 causes a segfault, it's out of my competences, I have no idea why.

Reply:
OK I'm still afraid it will not perform at its best, but what I did until now with various fixes functionally works fine. Now there is only one xdebug_hash lookup per function call, and only one previous stack item fetch (in the current stack). I finally understood why and how Fiber function calls did mess up the trace output order, I could fix the correlated unit test and documented it.